tasks/script.rb

```diff
@@ -5,25 +5,137 @@
 require 'json'
 require 'open3'
 require 'tempfile'
+require 'pathname'
+require 'fileutils'
 
-def command(command, arguments = [])
-  stdout, stderr, p = Open3.capture3(command, *arguments)
+module WinHelpers
+  class << self
+    PS_ARGS = %w[
+      -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass
+    ].freeze
+
+    def windows?
+      !!File::ALT_SEPARATOR
+    end
+
+    def powershell_script?(extension)
+      extension.casecmp('.ps1').zero?
+    end
+
+    def run_script(arguments, script_path)
+      mapped_args = arguments.map { |a|
+        "$invokeArgs.ArgumentList += @'\n#{a}\n'@"
+      }.join("\n")
+      <<-PS
+    $invokeArgs = @{
+      ScriptBlock = (Get-Command "#{script_path}").ScriptBlock
+      ArgumentList = @()
+    }
+    #{mapped_args}
+
+    try
+    {
+      Invoke-Command @invokeArgs
+    }
+    catch
+    {
+      Write-Error $_.Exception
+      exit 1
+    }
+    PS
+    end
+
+    def process_from_extension(path)
+      case Pathname(path).extname.downcase
+      when '.rb'
+        [
+          'ruby.exe',
```

Comment:
Will this be puppets ruby?

Reply:
I'm ok not officially supporting this for now.

Reply:
Oddly Ruby doesn't have a built-in mechanism for resolving the calling binary.

That said... *if* we know this runs on top of the Puppet Ruby interpreter (and we know that we have FFI available as a dependency), this code at https://stackoverflow.com/a/25929455 will work (I would make some minor mods to what's being done here, but the premise is correct):

```
[1] pry(main)> require 'ffi'
=> true
[2] pry(main)> module Helpers
[2] pry(main)*   extend FFI::Library
[2] pry(main)*
[2] pry(main)*   ffi_lib :kernel32
[2] pry(main)*
[2] pry(main)*   typedef :uintptr_t, :hmodule
[2] pry(main)*   typedef :ulong, :dword
[2] pry(main)*
[2] pry(main)*   attach_function :GetModuleHandle, :GetModuleHandleA, [:string], :hmodule
[2] pry(main)*   attach_function :GetModuleFileName, :GetModuleFileNameA, [:hmodule, :pointer, :dword], :dword
[2] pry(main)*
[2] pry(main)*   def self.actualRubyExecutable
[2] pry(main)*     processHandle = GetModuleHandle nil
[2] pry(main)*
[2] pry(main)*     # There is a potential issue if the ruby executable path is
[2] pry(main)*     # longer than 999 chars.
[2] pry(main)*     rubyPath = FFI::MemoryPointer.new 1000
[2] pry(main)*     rubyPathSize = GetModuleFileName processHandle, rubyPath, 999
[2] pry(main)*
[2] pry(main)*     rubyPath.read_string rubyPathSize
[2] pry(main)*   end
[2] pry(main)* end
=> :actualRubyExecutable
[3] pry(main)> Helpers.actualRubyExecutable
=> "C:\\tools\\ruby25\\bin\\ruby.exe"
```

I've noticed that there is already a `WinHelpers` class that's being called in this shim - but not sure where that code lives? Might be useful to land this there, depending on how this ships, etc.

Reply:
For some context, In order to support the local transport on windows bolt runners we pulled some powershell specific code into template functions https://github.com/puppetlabs/bolt/blob/master/lib/bolt/transport/powershell.rb which allowed us to share between the local and winrm transports. My thought was to apply those in this case as well. I chose to keep the bolt_shim::script implementation a single file to support older agents that did not allow multi file tasks.

I tested running the following ruby script on a windows 10 agent:
test.rb:
```
puts ARGV[0]
puts 'hi'
```
Invocation:
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt/Boltdir$ bolt script run test.rb 'im a param' -n hcbig1skqbpx759.delivery.puppetlabs.net --transport pcp
Started on hcbig1skqbpx759.delivery.puppetlabs.net...
Finished on hcbig1skqbpx759.delivery.puppetlabs.net:
  STDOUT:
    im a param
    hi
Successful on 1 node: hcbig1skqbpx759.delivery.puppetlabs.net
Ran on 1 node in 1.45 seconds
```
Feel free to test with them. The only setup I did on master was to clone this repo in `/etc/puppetlabs/code/environments/production/modules/bolt_shim` and checkout this PR. The windows node was simply set up with the `https://r8hrv1pfr6dk15b.delivery.puppetlabs.net:8140/packages/current/install.ps1` script

master:
`r8hrv1pfr6dk15b.delivery.puppetlabs.net`
Windows agent:
`hcbig1skqbpx759.delivery.puppetlabs.net`

Reply:
Per in-person discussion, we're going to raise an error on executing Ruby files on Windows for now, given there are likely no active users of this functionality.